test/envConfig.test.ts

```diff
@@ -49,3 +58,109 @@ describe('createEnvConfigContent', () => {
     expect(result).toMatchSnapshot();
   });
 });
+
+describe('envConfig', () => {
+  const FIXTURES_PATH = path.join(__dirname, 'fixtures');
+
+  describe('devServer', () => {
+    let server: ViteDevServer;
+    const env = process.env;
+
+    beforeEach(() => {
+      jest.resetModules();
+      process.env = { ...env };
+    });
+
+    afterEach(async () => {
+      process.env = env;
+      await server.close();
+    });
+
+    it('serves /index.html that includes envConfig.js script', async () => {
+      expect.assertions(2);
+
+      // given
+      server = await createServer({
+        root: FIXTURES_PATH,
+        plugins: [envConfig()],
+        server: {
+          port: 3000,
+        },
+      });
+      await server.listen();
+
+      // when
+      const response = await request('http://localhost:3000').get('/index.html');
+
+      // then
+      expect(response.status).toBe(200);
+      expect(response.text).toContain('<script src="/env-config.js"></script>');
+    });
+
+    it('serves /env-config.js', async () => {
+      expect.assertions(2);
+
+      // given
+      const variables = ['BACKEND_URL'];
+      process.env.BACKEND_URL = 'http://localhost:4000';
+
+      server = await createServer({
+        root: FIXTURES_PATH,
+        plugins: [envConfig({ variables })],
+        server: {
+          port: 3000,
+        },
+      });
+      await server.listen();
+
+      // when
+      const response = await request('http://localhost:3000').get('/env-config.js');
+
+      // then
+      expect(response.status).toBe(200);
+      expect(response.text).toContain(`window.env['BACKEND_URL']='${process.env.BACKEND_URL}';`);
+    });
+  });
+
+  describe('build', () => {
+    beforeEach(() => {
+      const distPath = path.join(FIXTURES_PATH, 'dist');
+      if (fs.existsSync(distPath)) {
+        fs.rmSync(distPath, { recursive: true });
+      }
+    });
+
+    it('modifies index.html to include envConfig.js script', async () => {
+      expect.assertions(1);
+
+      // when
+      await build({
+        logLevel: 'silent',
+        root: FIXTURES_PATH,
+        plugins: [envConfig()],
+      });
+      const indexHtml = await readFile(path.join(FIXTURES_PATH, 'dist/index.html'));
+
+      // then
+      expect(indexHtml.toString()).toContain('<script src="/env-config.js"></script>');
+    });
+
+    it('creates env-config.template.js containing variables and placeholders', async () => {
+      expect.assertions(1);
+
+      // given
+      const variables = ['BACKEND_URL'];
```

Comment:
Maybe it would be nice to move this up to not wirte it again

Reply:
There are tests with other `variables` arrays and not all tests use variables, so I don't think this is necessary.